lib/iris/cube.py

```diff
@@ -758,6 +767,231 @@ def _is_single_item(testee):
+                self.globals.update(deepcopy(combined.globals))
+                self.locals.update(deepcopy(combined.locals))
+            else:
+                # Treat any arbitrary single input value as a mapping (dict), and
+                # update from it.
+                self.update(dict(deepcopy(combined)))
+
+    #
+    # Ensure that the stored local/global dictionaries are "LimitedAttributeDicts".
+    #
+    @staticmethod
+    def _normalise_attrs(
+        attributes: Optional[Mapping],
+    ) -> LimitedAttributeDict:
+        # Convert an input attributes arg into a standard form.
+        # N.B. content is always a LimitedAttributeDict, and a deep copy of input.
+        # Allow arg of None, etc.
+        if not attributes:
+            attributes = {}
+        else:
+            attributes = deepcopy(attributes)
+
+        # Ensure the expected mapping type.
+        attributes = LimitedAttributeDict(attributes)
+        return attributes
+
+    @property
+    def locals(self):
+        return self._locals
+
+    @locals.setter
+    def locals(self, attributes):
+        self._locals = self._normalise_attrs(attributes)
+
+    @property
+    def globals(self):
+        return self._globals
+
+    @globals.setter
+    def globals(self, attributes):
+        self._globals = self._normalise_attrs(attributes)
+
+    #
+    # Provide a serialisation interface
+    #
+    def __getstate__(self):
+        return (self.locals, self.globals)
+
+    def __setstate__(self, state):
+        self.locals, self.globals = state
+
+    #
+    # Support simple comparison, even when contents are arrays.
+    #
+    def __eq__(self, other):
+        # For equality, require both globals + locals to match
+        other = CubeAttrsDict(other)
+        result = self.locals == other.locals and self.globals == other.globals
+        return result
+
+    def __ne__(self, other):
+        return not self == other
+
+    #
+    # Provide a copy method, as for 'dict', but *not* provided by MutableMapping
+    #
+    def copy(self):
+        """
+        Return a copy.
+
+        Implemented with deep copying, consistent with general Iris usage.
+
+        """
+        return CubeAttrsDict(self)
+
+    #
+    # The remaining methods are sufficient to generate a complete standard Mapping
+    # API -- see docs for :class:`collections.abc.MutableMapping`.
+    #
+
+    def __iter__(self):
+        # Ordering: all global keys, then all local ones, but omitting duplicates.
+        # NOTE: this means that in the "summary" view, attributes present in both
+        # locals+globals are listed first, amongst the globals, even though they appear
+        # with the *value* from locals.
+        # Otherwise follows order of insertion, as is normal for dicts.
+        return itertools.chain(
+            self.globals.keys(),
+            (x for x in self.locals.keys() if x not in self.globals),
+        )
```

Comment:
I don't think this is a good replacement : it produces the correct set of attribute names, but not in the desired _order_.  
I think it's important that the order is that of the globals, followed by locals-where-different, as stated.  Neither "locals | globals" or "globals | locals" reproduces that correctly.

I also think we don't need a special 'combined' method, since `dict(self)` provides exactly that, based on the iteration order as defined here.  That is actually used in the `__str__` function, but by contrast `len` is based directly on `iter`, since using a combined dictionary there would be making a copy every time.